client/cli/src/params/shared_params.rs

```diff
@@ -39,51 +38,47 @@ pub struct SharedParams {
 	pub dev: bool,
 
 	/// Specify custom base path.
-	#[structopt(long = "base-path", short = "d", value_name = "PATH", parse(from_os_str))]
+	// NOTE: this is an option so implementations can set their own defaults
+	#[structopt(
+		long = "base-path",
+		short = "d",
+		value_name = "PATH",
+		parse(from_os_str)
+	)]
 	pub base_path: Option<PathBuf>,
 
 	/// Sets a custom logging filter. Syntax is <target>=<level>, e.g. -lsync=debug.
 	///
 	/// Log levels (least to most verbose) are error, warn, info, debug, and trace.
 	/// By default, all targets log `info`. The global log level can be set with -l<level>.
+	// NOTE: this is an option so implementations can set their own defaults
 	#[structopt(short = "l", long = "log", value_name = "LOG_PATTERN")]
 	pub log: Option<String>,
 }
 
 impl SharedParams {
-	/// Load spec to `Configuration` from `SharedParams` and spec factory.
-	pub fn update_config<'a, F>(
-		&self,
-		mut config: &'a mut Configuration,
-		spec_factory: F,
-		version: &VersionInfo,
-	) -> error::Result<&'a dyn ChainSpec> where
-		F: FnOnce(&str) -> Result<Box<dyn ChainSpec>, String>,
-	{
-		let chain_key = match self.chain {
+	/// Specify custom base path.
+	pub fn base_path(&self) -> Result<Option<PathBuf>> {
+		Ok(self.base_path.clone())
+	}
+
+	/// Specify the development chain.
+	pub fn is_dev(&self) -> Result<bool> {
+		Ok(self.dev)
+	}
+
+	/// Get the chain spec for the parameters provided
+	pub fn chain_id(&self, is_dev: bool) -> Result<String> {
+		Ok(match self.chain {
 			Some(ref chain) => chain.clone(),
-			None => if self.dev { "dev".into() } else { "".into() }
-		};
-		let spec = spec_factory(&chain_key)?;
-		config.network.boot_nodes = spec.boot_nodes().to_vec();
-		config.telemetry_endpoints = spec.telemetry_endpoints().clone();
-
-		config.chain_spec = Some(spec);
-
-		if config.config_dir.is_none() {
-			config.config_dir = Some(base_path(self, version));
-		}
-
-		if config.database.is_none() {
-			config.database = Some(DatabaseConfig::Path {
-				path: config
-					.in_chain_config_dir(DEFAULT_DB_CONFIG_PATH)
-					.expect("We provided a base_path/config_dir."),
-				cache_size: None,
-			});
-		}
-
-		Ok(config.expect_chain_spec())
+			None => {
+				if is_dev {
+					"dev".into()
```

Comment:
It is rather weird that this returns `dev`. I know that you can overwrite this function, but still weird. For polkadot it is for example `polkadot-dev` and `kusama-dev`.

Reply:
Loading the spec probably requires more refactoring. It was already stringly-typed like that. In polkadot you can see that I actually removed an entire enum and a 2-step process (from && load) because it was unnecessary. I think we can open an issue for this.